sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1442,6 +1442,12 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
       pushDownPredicate(filter, u.child) { predicate =>
         u.withNewChildren(Seq(Filter(predicate, u.child)))
       }
+
+    // Push down filter predicates in case filter having child as TypedFilter.
+    // In this scenario inorder to push the filter predicates there is need to
+    // to push Filter beneath the TypedFilter.
+    case Filter(condition, typeFilter @ TypedFilter(_, _, _, _, _)) =>
+      typeFilter.copy(child = Filter(condition, typeFilter.child))
```

Comment:
The only thing makes me worried is `TypedFilter` takes an arbitrary function that can depend on, e.g, number of records being filtered or any sort of global variable for the task. Once the order is switched, the logic can be broken. e.g.)

```scala
object RecordFilteredPerTask {
    var count = 0
}

df.filter({ v =>
  RecordFilteredPerTask.count += 1
  RecordFilteredPerTask.count > 5
})
```

Reply:
@HyukjinKwon  - Thank you for reviewing this PR.

Yes that makes sense for some function that are selecting the first n records on each task and applying the filter on top of that.If the filter sequence change the output will be different.

In my scenario which is having typeFilter on checking the value that exist in the bloomFilter. I have to add as this rule as the part of "SparkSession.experimental.extraOptimizations = Seq()" then.

Any other suggestion on this.How to take on these kind of scenario.

Reply:
Hmm, it looks like we don't well document if typed filter function can be non-deterministic or not. If we look at UDF, it is deterministic by default, though you can ask it to be non-deterministic.

So that's said, I suppose the typed filter function is deterministic, though we don't provide a way to make it as non-deterministic.

Reply:
To be safer here, we may exclude the possibility that users use non-deterministic typed filter function, actually.

Reply:
BTW, looks like typed filter is separated from normal filter. So it should be easier to adjust filter, typed filter operator in queries to make filter pushdown-able even it is not optimized?

Reply:
**BTW, looks like typed filter is separated from normal filter. So it should be easier to adjust filter, typed filter operator in queries to make filter pushdown-able even it is not optimized?**

This is what you mean to say here 

Like this in the query itself df.filter("id=1").filter(SomeTypeFilter) instead of df.filter(SomeTypeFilter).filter("id=1").

For this we need to tell user that used typeFilter after the normal filter, since normal filter might partitioned filter  and improve the the performance of the query. I thought if we can do something in the optimizer rule to handle such scenarios